### PR TITLE
feat: Implement Nonce Handling for Proposals

### PR DESCRIPTION
**Pull request by jyotirmoydotdev**
## ****Signature Verification****

- **Old Contract:** The old contract does not use any signature verification mechanism. It relies solely on the ownership of the NFT to approve proposals.
- **Updated Contract:** Implements a signature verification mechanism using ECDSA. This ensures that approvals are based on a cryptographic signature from the owner of the NFT, providing an additional layer of security.

```solidity
function verifySignature(uint8 proposalId, uint8 tokenId, bytes memory signature) external view returns (bool) {
    bytes32 messageHash = constructMessageHash(proposalId, tokenId);
    address signer = ECDSA.recover(messageHash, signature);
    return signer == IERC721(memberToken).ownerOf(tokenId);
}
```

## **Nonce Implementation**

- **Old Contract:** Lacks a nonce mechanism, making it susceptible to replay attacks.
- **Updated Contract:** Introduces a nonce for each proposal. This prevents replay attacks by ensuring that a signature is valid only for a specific proposal and cannot be reused for other transactions.

```solidity
/// @notice Counter to track the nonce for each proposal
uint256 public nonce;
```

## **Enhanced Proposal Structure**

- **Old Contract:** Uses a relatively simple structure for proposals without a nonce and does not include nonces in the verification process.
- **Updated Contract:** Enhances the proposal structure by including a nonce. This improves the uniqueness and security of each proposal, mitigating potential attacks that exploit the absence of a nonce.

```solidity
struct Proposal {
    // ... other fields
    uint256 nonce;
}
```

## **Security Improvements and Attack Prevention**

1. **Replay Attacks:** The introduction of nonces in the Updated contract prevents replay attacks by ensuring that each signature is valid only for a specific proposal, reducing the risk of malicious re-execution.
2. **Signature Verification:** The use of ECDSA for signature verification enhances the security of the approval process, making it more resilient against attacks attempting to forge approvals.
3. **Improved Proposal Structure:** The inclusion of a nonce in the proposal structure enhances the uniqueness of each proposal, preventing potential attacks that might exploit the absence of nonces.

### CODE REVIEW

File: src/Common.sol [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 pragma solidity 0.8.19;
 
-
+import {ECDSA} from "openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
```
Comment (xhad) [inline src/Common.sol]: kindly format the syntax as the other imports { MODULE }

Comment (xhad) [reply src/Common.sol]: also, please move the ECDSA import in the following order:

```
import { IERC20 } from "openzeppelin-contracts/contracts/interfaces/IERC20.sol";
import { IERC721 } from "openzeppelin-contracts/contracts/interfaces/IERC721.sol";
import { ECDSA } from "openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
import { Context } from "openzeppelin-contracts/contracts/utils/Context.sol";
import { SafeERC20 } from "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
```

**Review by xhad — CHANGES_REQUESTED** [review]
This looks really nice, just kindly move / change the ECDSA import in Common.sol

File: src/Common.sol (hunk 1 above)
Comment (jyotirmoydotdev) [reply src/Common.sol]: Done

**Pull request merged**